File: python/5.py
import unittest
from typing import List


class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return s

        dp: List[List[bool]] = list(
            list(False for _ in range(len(s))) for _ in range(len(s)))

        resultL = 0
        resultR = 0
        for l in range(1, len(s) + 1):
            for i in range(len(s) - l + 1):
                j = i + l - 1
                dp[i][j] = s[i] == s[j] and (l < 3 or dp[i + 1][j - 1])
                if dp[i][j] and l > max:
                    resultL = i
                    resultR = j

        return s[resultL:resultR + 1]


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual("bab", solution.longestPalindrome("babad"))
        self.assertEqual("bb", solution.longestPalindrome("cbbd"))

File: typescript/5.ts
function longestPalindrome(s: string): string {
    if (!s.length) return s;

    const dp: boolean[][] = new Array(s.length).fill(false).map(
        () => new Array(s.length).fill(false)
    );

    let resultL = 0;
    let resultR = 0;
    for (let l = 1; l <= s.length; l++) {
        for (let i = 0; i + l <= s.length; i++) {
            const j = i + l - 1;
            dp[i][j] = s[i] === s[j] && (l <= 2 || dp[i + 1][j - 1]);
            if (dp[i][j]) {
                resultL = i;
                resultR = j;
            }
        }
    }

    return s.slice(resultL, resultR + 1);
};

test('5', () => {
    expect(longestPalindrome('babad')).toBe('aba');
    expect(longestPalindrome('cbbd')).toBe('bb');
});